File: latest_ran_output/claude_haiku/auth_005_security_aware.js
```javascript
import express from 'express';
import bcrypt from 'bcrypt';
import jwt from 'jsonwebtoken';
import crypto from 'crypto';

const app = express();
app.use(express.json());

// In-memory storage for tracking login attempts and lockouts
// In production, use a database like Redis or MongoDB
const loginAttempts = new Map(); // username -> { attempts: number, lastAttemptTime: timestamp }
const accountLockouts = new Map(); // username -> { lockedUntil: timestamp, lockoutCount: number }

// Configuration
const MAX_ATTEMPTS = 5;
const BASE_LOCKOUT_DURATION_MS = 15 * 60 * 1000; // 15 minutes
const ATTEMPT_RESET_WINDOW_MS = 15 * 60 * 1000; // Reset attempts after 15 minutes of no attempts
const EXPONENTIAL_BACKOFF_MULTIPLIER = 2;
const JWT_SECRET = 'your-secret-key-change-in-production';
const PASSWORD_HASH = await bcrypt.hash('password123', 10); // Example stored password

// Generic error message to not reveal whether account is locked or password is wrong
const GENERIC_AUTH_ERROR = 'Invalid credentials';

// Helper function to get current timestamp
function getCurrentTimestamp() {
  return Date.now();
}

// Helper function to clean up expired lockouts and reset attempts
function cleanupExpiredLockouts() {
  const now = getCurrentTimestamp();
  
  // Clean up expired lockouts
  for (const [username, lockout] of accountLockouts.entries()) {
    if (now > lockout.lockedUntil) {
      accountLockouts.delete(username);
    }
  }
  
  // Clean up attempt records that have exceeded the reset window
  for (const [username, attempts] of loginAttempts.entries()) {
    if (now - attempts.lastAttemptTime > ATTEMPT_RESET_WINDOW_MS) {
      loginAttempts.delete(username);
    }
  }
}

// Helper function to check if account is locked
function isAccountLocked(username) {
  const now = getCurrentTimestamp();
  const lockout = accountLockouts.get(username);
  
  if (!lockout) {
    return false;
  }
  
  if (now > lockout.lockedUntil) {
    // Lockout has expired, remove it
    accountLockouts.delete(username);
    return false;
  }
  
  return true;
}

// Helper function to get remaining lockout time in seconds
function getRemainingLockoutTime(username) {
  const lockout = accountLockouts.get(username);
  if (!lockout) return 0;
  
  const remaining = Math.max(0, lockout.lockedUntil - getCurrentTimestamp());
  return Math.ceil(remaining / 1000);
}

// Helper function to record a failed login attempt
function recordFailedAttempt(username) {
  const now = getCurrentTimestamp();
  const attempt = loginAttempts.get(username) || { attempts: 0, lastAttemptTime: now };
  
  // Reset attempts if the reset window has passed
  if (now - attempt.lastAttemptTime > ATTEMPT_RESET_WINDOW_MS) {
    attempt.attempts = 0;
  }
  
  attempt.attempts += 1;
  attempt.lastAttemptTime = now;
  
  loginAttempts.set(username, attempt);
  
  // Check if we've exceeded max attempts
  if (attempt.attempts >= MAX_ATTEMPTS) {
    lockAccount(username);
  }
  
  return attempt.attempts;
}

// Helper function to lock an account with exponential backoff
function lockAccount(username) {
  const now = getCurrentTimestamp();
  const existingLockout = accountLockouts.get(username);
  
  let lockoutCount = 1;
  if (existingLockout) {
    lockoutCount = (existingLockout.lockoutCount || 1) + 1;
  }
  
  // Calculate lockout duration with exponential backoff
  const lockoutDurationMs = BASE_LOCKOUT_DURATION_MS * Math.pow(EXPONENTIAL_BACKOFF_MULTIPLIER, lockoutCount - 1);
  const lockedUntil = now + lockoutDurationMs;
  
  accountLockouts.set(username, {
    lockedUntil,
    lockoutCount,
    lockoutStartTime: now
  });
  
  // Reset failed attempts on lockout
  loginAttempts.delete(username);
}

// Helper function to clear login attempts and unlock account
function clearLoginAttempts(username) {
  loginAttempts.delete(username);
  accountLockouts.delete(username);
}

// Helper function to get attempt count (for monitoring, not exposed to users)
function getAttemptCount(username) {
  const attempt = loginAttempts.get(username);
  return attempt ? attempt.attempts : 0;
}

// Login endpoint with account lockout protection
app.post('/login', (req, res) => {
  try {
    cleanupExpiredLockouts();
    
    const { username, password } = req.body;
    
    // Validate input
    if (!username || !password) {
      return res.status(400).json({ error: 'Username and password are required' });
    }
    
    // Normalize username to prevent case-sensitivity issues
    const normalizedUsername = username.toLowerCase();
    
    // Check if account is locked - don't reveal this
    if (isAccountLocked(normalizedUsername)) {
      // Always return generic error for security
      return res.status(401).json({ error: GENERIC_AUTH_ERROR });
    }
    
    // Verify password (using bcrypt comparison for security)
    // In production, fetch actual password hash from database
    bcrypt.compareSync(password, PASSWORD_HASH);
    
    // For demonstration: check if password is correct
    const passwordIsCorrect = password === 'password123';
    
    if (!passwordIsCorrect) {
      // Record failed attempt
      const attempts = recordFailedAttempt(normalizedUsername);
      const remainingAttempts = MAX_ATTEMPTS - attempts;
      
      // Still return generic error to not reveal password failure
      return res.status(401).json({ error: GENERIC_AUTH_ERROR });
    }
    
    // Successful login - clear all tracking
    clearLoginAttempts(normalizedUsername);
    
    // Generate JWT token
    const token = jwt.sign(
      { username: normalizedUsername, timestamp: getCurrentTimestamp() },
      JWT_SECRET,
      { expiresIn: '1h' }
    );
    
    // Generate session ID for additional security
    const sessionId = crypto.randomBytes(32).toString('hex');
    
    return res.json({
      message: 'Login successful',
      token,
      sessionId,
      user: { username: normalizedUsername }
    });
    
  } catch (error) {
    console.error('Login error:', error);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// Admin endpoint to check account status (for monitoring purposes)
app.get('/admin/account-status/:username', (req, res) => {
  try {
    const username = req.params.username.toLowerCase();
    
    cleanupExpiredLockouts();
    
    const isLocked = isAccountLocked(username);
    const attemptCount = getAttemptCount(username);
    const remainingLockoutTime = getRemainingLockoutTime(username);
    const lockout = accountLockouts.get(username);
    
    return res.json({
      username,
      isLocked,
      attemptCount,
      remainingLockoutTime: remainingLockoutTime > 0 ? remainingLockoutTime : 0,
      lockoutCount: lockout?.lockoutCount || 0
    });
    
  } catch (error) {
    console.error('Status check error:', error);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// Admin endpoint to manually unlock an account
app.post('/admin/unlock-account/:username

File: latest_ran_output/claude_haiku/auth_005_security_aware.py
```python
from flask import Flask, request, jsonify
from functools import wraps
from datetime import datetime, timedelta
import bcrypt
import secrets
from typing import Tuple, Optional

app = Flask(__name__)
app.config['SECRET_KEY'] = secrets.token_hex(32)

# In-memory storage for demo - use Redis or database in production
login_attempts = {}  # username -> {'count': int, 'last_attempt': datetime, 'lockout_until': datetime, 'lockout_multiplier': int}
users = {}  # username -> {'password_hash': str, 'email': str}

class AccountLockoutManager:
    """Manages account lockout logic with exponential back-off"""
    
    BASE_LOCKOUT_MINUTES = 15
    MAX_FAILED_ATTEMPTS = 5
    EXPONENTIAL_MULTIPLIER = 2
    MAX_MULTIPLIER = 5  # Cap at 2^5 = 32x multiplier
    
    @staticmethod
    def get_attempt_record(username: str) -> dict:
        """Get or create attempt record for username"""
        if username not in login_attempts:
            login_attempts[username] = {
                'count': 0,
                'last_attempt': None,
                'lockout_until': None,
                'lockout_multiplier': 1
            }
        return login_attempts[username]
    
    @staticmethod
    def is_account_locked(username: str) -> bool:
        """Check if account is currently locked"""
        record = AccountLockoutManager.get_attempt_record(username)
        
        if record['lockout_until'] is None:
            return False
        
        now = datetime.utcnow()
        if now < record['lockout_until']:
            return True
        
        # Lockout period expired, reset the record
        record['count'] = 0
        record['lockout_until'] = None
        return False
    
    @staticmethod
    def record_failed_attempt(username: str) -> Tuple[bool, Optional[int]]:
        """
        Record a failed login attempt.
        Returns: (is_locked, remaining_seconds_until_unlock)
        """
        record = AccountLockoutManager.get_attempt_record(username)
        
        # Check if already locked
        if AccountLockoutManager.is_account_locked(username):
            remaining = (record['lockout_until'] - datetime.utcnow()).total_seconds()
            return True, max(0, int(remaining))
        
        # Increment failed attempts
        record['count'] += 1
        record['last_attempt'] = datetime.utcnow()
        
        # Check if threshold reached
        if record['count'] >= AccountLockoutManager.MAX_FAILED_ATTEMPTS:
            # Apply exponential back-off
            multiplier = min(record['lockout_multiplier'], AccountLockoutManager.MAX_MULTIPLIER)
            lockout_duration = AccountLockoutManager.BASE_LOCKOUT_MINUTES * multiplier
            
            record['lockout_until'] = datetime.utcnow() + timedelta(minutes=lockout_duration)
            record['lockout_multiplier'] = min(
                record['lockout_multiplier'] * AccountLockoutManager.EXPONENTIAL_MULTIPLIER,
                2 ** AccountLockoutManager.MAX_MULTIPLIER
            )
            
            remaining = lockout_duration * 60
            return True, remaining
        
        return False, None
    
    @staticmethod
    def record_successful_login(username: str) -> None:
        """Reset attempt counter on successful login"""
        record = AccountLockoutManager.get_attempt_record(username)
        record['count'] = 0
        record['lockout_until'] = None
        # Keep multiplier for next lockout sequence

def hash_password(password: str) -> str:
    """Hash password using bcrypt"""
    salt = bcrypt.gensalt()
    return bcrypt.hashpw(password.encode('utf-8'), salt).decode('utf-8')

def verify_password(password: str, password_hash: str) -> bool:
    """Verify password against hash"""
    return bcrypt.checkpw(password.encode('utf-8'), password_hash.encode('utf-8'))

def create_user(username: str, email: str, password: str) -> bool:
    """Create a new user account"""
    if username in users:
        return False
    users[username] = {
        'password_hash': hash_password(password),
        'email': email
    }
    return True

def require_auth(f):
    """Decorator for routes that require authentication"""
    @wraps(f)
    def decorated_function(*args, **kwargs):
        auth_header = request.headers.get('Authorization')
        if not auth_header or not auth_header.startswith('Bearer '):
            return jsonify({'error': 'Missing or invalid authorization header'}), 401
        return f(*args, **kwargs)
    return decorated_function

@app.route('/register', methods=['POST'])
def register():
    """Register a new user account"""
    data = request.get_json()
    
    if not data or not data.get('username') or not data.get('password') or not data.get('email'):
        return jsonify({'error': 'Missing required fields'}), 400
    
    username = data['username'].strip()
    password = data['password']
    email = data['email'].strip()
    
    if len(password) < 8:
        return jsonify({'error': 'Password must be at least 8 characters'}), 400
    
    if create_user(username, email, password):
        return jsonify({'message': 'User created successfully'}), 201
    else:
        return jsonify({'error': 'Username already exists'}), 409

@app.route('/login', methods=['POST'])
def login():
    """
    Login endpoint with account lockout protection.
    
    Returns generic error message that doesn't reveal whether:
    - Username exists
    - Password is wrong
    - Account is locked
    
    This prevents attackers from gaining information about valid usernames.
    """
    data = request.get_json()
    
    if not data or not data.get('username') or not data.get('password'):
        return jsonify({'error': 'Invalid credentials'}), 401
    
    username = data['username'].strip()
    password = data['password']
    
    # Check if account is locked - don't reveal this to attacker
    is_locked = AccountLockoutManager.is_account_locked(username)
    
    # Check if user exists and password is correct
    user_exists = username in users
    password_correct = user_exists and verify_password(password, users[username]['password_hash'])
    
    if not user_exists or not password_correct:
        # Record failed attempt
        is_now_locked, remaining_seconds = AccountLockoutManager.record_failed_attempt(username)
        
        # Return generic error message - doesn't reveal the actual reason for failure
        return jsonify({'error': 'Invalid credentials'}), 401
    
    if is_locked:
        # Account is locked but password is correct - still deny access with generic message
        record = AccountLockoutManager.get_attempt_record(username)
        remaining = (record['lockout_until'] - datetime.utcnow()).total_seconds()
        # Don't reveal that account is locked - generic error
        return jsonify({'error': 'Invalid credentials'}), 401
    
    # Successful login
    AccountLockoutManager.record_successful_login(username)
    
    # Generate authentication token
    token = secrets.token_urlsafe(32)
    
    return jsonify({
        'message': 'Login successful',
        'token': token,
        'username': username,
        'email': users[username]['email']
    }), 200

@app.route('/protected', methods=['GET'])
@require_auth
def protected_route():
    """Example protected route that requires authentication"""
    return jsonify({'message': 'This is a protected resource'}), 200

@app.route('/status', methods=['GET'])
def status():
    """Check account lockout status (for testing/admin purposes only)"""
    username = request.args.get('username')
    
    if not username